app/Platform/Services/UserAgentService.php

```diff
@@ -163,4 +164,101 @@ private function trimOperatingSystem(string $os): string
     {
         return trim(strtok($os, ';'));
     }
+
+    public static function versionCompare(string $versionOne, string $versionTwo): int
+    {
+        $versionOneParts = UserAgentService::splitVersion($versionOne);
+        $versionOnePartCount = count($versionOneParts);
+        $versionTwoParts = UserAgentService::splitVersion($versionTwo);
+        $versionTwoPartCount = count($versionTwoParts);
+
+        $index = 0;
+        while (true) {
+            if ($index === $versionOnePartCount) {
+                return ($index === $versionTwoPartCount) ? 0 : -1;
+            }
+
+            if ($index === $versionTwoPartCount) {
+                return 1;
+            }
+
+            $versionOnePart = $versionOneParts[$index];
+            $versionTwoPart = $versionTwoParts[$index];
+            $diff = strcmp($versionOnePart, $versionTwoPart);
+            if ($diff !== 0) {
+                $versionOneNumeric = ctype_digit($versionOnePart);
+                $versionTwoNumeric = ctype_digit($versionTwoPart);
+                if ($versionOneNumeric) {
+                    if ($versionTwoNumeric) {
+                        // both parts are fully numeric. compare them numerically
+                        $versionOneNumber = (int) $versionOnePart;
+                        $versionTwoNumber = (int) $versionTwoPart;
+                        if ($versionOneNumber < $versionTwoNumber) {
+                            return -1;
+                        } elseif ($versionOneNumber > $versionTwoNumber) {
+                            return 1;
+                        } else {
+                            return 0;
+                        }
+                    }
+
+                    // 1.0.1 > 1.0-dirty
+                    return 1;
+                } elseif ($versionTwoNumeric) {
+                    // 1.0-dirty < 1.0.1
+                    return -1;
+                }
+
+                // neither part is fully numeric. compare them as strings
+                return ($diff < 0) ? -1 : 1;
+            }
+
+            $index++;
+        }
+    }
+
+    private static function splitVersion(string $version): array
```

Comment:
What happens if I release version 1.0a and it's followed by 1.0B?

Reply:
`1.0a` is converted to `1.0.1` and `1.0b` is converted to `1.0.2` by this code:
https://github.com/RetroAchievements/RAWeb/pull/2822/files#diff-0e7cd5df70a77fe7e572b4a371f1cf4390006cf2cac2c57378ab285fbbe4c679R248

Reply:
`1.0b` is handled correctly, but the algorithm is case sensitive, eg:
```php
public function testSampleAgent(): void
{
    $this->assertEquals(1, UserAgentService::versionCompare('1.0B', '1.0a')); // FAILS
}
```

Reply:
fixed